File: movecraft-core/src/main/java/net/countercraft/movecraft/utils/BlockNames.java
package net.countercraft.movecraft.utils;

import com.google.common.collect.ImmutableSet;
import net.countercraft.movecraft.api.MaterialDataPredicate;
import net.minecraft.server.v1_12_R1.Item;
import net.minecraft.server.v1_12_R1.ItemStack;
import org.bukkit.Material;
import org.bukkit.material.MaterialData;

public final class BlockNames {
    public static String properCase(final String text) {
        final char[] chars = text.toCharArray();
        boolean makeUpper = true;
        for (int ix = 0; ix < chars.length; ix++) {
            final char ch = makeUpper ? Character.toUpperCase(chars[ix]) : Character.toLowerCase(chars[ix]);
            makeUpper = !Character.isLetter(ch);
            chars[ix] = makeUpper ? ' ' : ch;
        }

        return new String(chars).replaceAll("\\s\\s+", " ").trim();
    }

    public static ImmutableSet<String> materialDataPredicateNames(final MaterialDataPredicate predicate) {
        final ImmutableSet.Builder<String> builder = ImmutableSet.builder();

        for (final Material material : predicate.allMaterials()) {
            builder.add(itemName(material));
        }
        for (final MaterialData materialDataPair : predicate.allMaterialDataPairs()) {
            builder.add(itemName(materialDataPair));
        }

        return builder.build();
    }

    private static String itemName(final Material material, final byte data, final boolean hasData) {
        String tmp = null;

        try {
            tmp = new ItemStack(Item.getById(material.getId()), 1, data).getName();
        } catch (final Exception ignored) { }

        if (tmp == null || tmp.isEmpty()) {
            tmp = material.name();
        }

        tmp = properCase(tmp);
        tmp = tmp.replaceAll("\\s(On|Off)$", "");
        if (!hasData && tmp.startsWith("White ")) tmp = tmp.substring(6);
        return tmp;
    }

    public static String itemName(final MaterialData materialData) {
        return itemName(materialData.getItemType(), materialData.getData(), true);
    }

    public static String itemName(final Material material, final byte data) {
        return itemName(material, data, true);
    }

    public static String itemName(final Material material) {
        return itemName(material, (byte) 0, false);
    }

    private static final ImmutableSet<Material> COLORED_MATERIALS = ImmutableSet
            .of(Material.WOOL, Material.CARPET, Material.STAINED_GLASS, Material.STAINED_GLASS_PANE,
                Material.STAINED_CLAY);

    public static ImmutableSet<String> itemNames(final Material blk) {
        // Wool, Carpet, Stained Glass, Glass Pane, Clay
        if (COLORED_MATERIALS.contains(blk)) {
            return ImmutableSet.of(itemName(blk));
        }

        final ImmutableSet.Builder<String> builder = ImmutableSet.builder();
        for (byte data = 0; data < 16; data++) {
            builder.add(itemName(blk, data, true));
        }
        return builder.build();
    }

    public static ImmutableSet<String> itemNames(final MaterialData blk) {
        return ImmutableSet.of(itemName(blk));
    }
}

File: movecraft-core/src/main/java/net/countercraft/movecraft/utils/BlockNames.scala
package net.countercraft.movecraft.utils

import com.google.common.collect.ImmutableSet
import com.alexknvl.shipcraft.MaterialDataPredicate
import net.minecraft.server.v1_12_R1.Item
import net.minecraft.server.v1_12_R1.ItemStack
import org.bukkit.Material
import org.bukkit.material.MaterialData

import scala.annotation.tailrec

object BlockNames {
  def properCase(text: String): String = {
    @tailrec def go(ix: Int, makeUpper: Boolean, chars: Array[Char]): Array[Char] =
      if (ix < chars.length) {
        val ch =
          if (makeUpper) Character.toUpperCase(chars(ix))
          else Character.toLowerCase(chars(ix))
        val isLetter = Character.isLetter(ch)
        chars(ix) = if (isLetter) ch else ' '
        go(ix + 1, !isLetter, chars)
      } else chars

    new String(go(0, makeUpper = true, text.toCharArray)).replaceAll("\\s\\s+", " ").trim
  }

  def materialDataPredicateNames(predicate: MaterialDataPredicate): Set[String] =
    predicate.allMaterials.map(itemName) ++ predicate.allMaterialDataPairs.map(itemName)

  private def itemName(material: Material, data: Byte, hasData: Boolean) = {
    var tmp =
      try new ItemStack(Item.getById(material.getId), 1, data).getName
      catch { case ignored: Exception => null }
    if (tmp == null || tmp.isEmpty) tmp = material.name
    tmp = properCase(tmp).replaceAll("\\s(On|Off)$", "")
    if (!hasData && tmp.startsWith("White "))
      tmp = tmp.substring(6)
    tmp
  }

  def itemName(materialData: MaterialData): String =
    itemName(materialData.getItemType, materialData.getData, hasData = true)
  def itemName(material: Material, data: Byte): String =
    itemName(material, data, hasData = true)
  def itemName(material: Material): String =
    itemName(material, 0.toByte, hasData = false)

  private val COLORED_MATERIALS = Set(
    Material.WOOL, Material.CARPET, Material.STAINED_GLASS,
    Material.STAINED_GLASS_PANE, Material.STAINED_CLAY)

  def itemNames(blk: Material): Set[String] = {
    // Wool, Carpet, Stained Glass, Glass Pane, Clay
    if (COLORED_MATERIALS.contains(blk)) Set(itemName(blk))
    else {
      val builder = Set.newBuilder[String]
      for (data <- 0 until 16) {
        builder.+=(itemName(blk, data.toByte, true))
      }
      builder.result()
    }
  }

  def itemNames(blk: MaterialData): Set[String] = Set(itemName(blk))
}